src/utils/retry.ts

```diff
@@ -2,21 +2,29 @@ type RetryOptions = {
   maxRetries?: number;
   initialDelay?: number;
   shouldRetry?: (error: Error) => boolean;
+  timeout?: number; // Total time in milliseconds before giving up
 };
 
 const delay = (ms: number): Promise<void> => new Promise(resolve => setTimeout(resolve, ms));
 
+const withTimeout = <T>(promise: Promise<T>, ms: number): Promise<T> => {
+  const timeoutPromise = new Promise<never>((_, reject) => {
+    setTimeout(() => reject(new Error(`Operation timed out after ${ms}ms`)), ms);
+  });
+  return Promise.race([promise, timeoutPromise]);
+};
+
```

Comment:
My only suggestion is to clear the timeouts if the error comes first! GC handles this properly; however, if we have too many retries, it potentially causes memory leakage.

```suggestion
const withTimeout = <T>(promise: Promise<T>, ms: number): Promise<T> => {
  let timeoutId: NodeJS.Timeout;
  
  const timeoutPromise = new Promise<never>((_, reject) => {
    timeoutId = setTimeout(() => reject(new Error(`Operation timed out after ${ms}ms`)), ms);
  });
  
  return Promise.race([
    promise.then(result => {
      clearTimeout(timeoutId);
      return result;
    }).catch(error => {
      clearTimeout(timeoutId);
      throw error;
    }),
    timeoutPromise
  ]);
};

```